### PR TITLE
[Breaking Change] Make Mesh-Level Boundary Conditions Usable without "User" flag (now without bitrot)

### PR DESCRIPTION
**Pull request by Yurlungur**
## NOTICE

This is my second attempt at #989 . The code has changed so much since my original implementation that it was easier to start from scratch rather than trying to merge with develop. Some notes:

- Block boundaries are used by forest of octrees. I now leave those alone.
- At @pgrete 's request, reflecting boundaries are no longer automatically registered. You may register them with the BC infrastructure however.
- This MR depends on #1176 . Please merge that one first. The diff will also be a lot easier to read after that change.

## API Breaking change

When registering user boundary functions, the mechanism changes from assigning the function pointer to an array. You now must call `ApplicationInput::RegisterBoundaryCondition` and give the custom boundary condition a name. This also eliminates the need for the `user` boundary flag in the input file, although that flag is implicitly still supported.

## PR Summary

A discussion on the new downstream code headed by @brryan @pdmullen indicated a desire to set mesh-level boundary conditions without using the `user` flag. Here I implement this feature. The changes I made are:
1. `ApplicationInput` now has a registration function where you register either a grid or particle boundary condition, with a function pointer (or a class in the case of swarm bounds) and a name.
2. ALL boundary conditions (except periodic) use this mechanism. Reflecting boundaries are, for example, registered in the constructor for `ApplicationInput` with the `"reflecting"` name.
3. `BoundaryFlag`s are dramatically simplified, indicating (essentially) only whether or not boundaries are periodic.
4. The mesh no longer uses flags to set function pointers. Instead, it just pulls the functions from `ApplicationInput`.
5. For (sort of) backwards compatibility, if you register a function without naming it, it's named "user."
6. Swarm boundaries are now decoupled from mesh boundaries. Although by default, they are set to the same value as the mesh.

The advantage of this is that you can now call (in your main function)
```C++
pman->app_input.RegisterBoundaryCondition(parthenon::BoundaryFace::inner_x1, "my_bc", &my_bc_func);
```
and then in the input file just use
```
<parthenon/mesh>
ix1_bc = my_bc
```

I think the refactored code is also significantly cleaner and easier to read. Also I put a link to the parthenon paper in the readme.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [x] Change is breaking (API, behavior, ...)
  - [x] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [x] PR is marked as breaking
  - [x] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [x] Docs build
- [x] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Comment by Yurlungur** [comment]
Ping @pgrete, @brryan , @bprather and @adamdempsey90 you guys asked for this. Please review when you get the chance. :)

File: doc/sphinx/src/boundary_conditions.rst [hunk 1]
```diff
@@ -1,4 +1,4 @@
-.. _sphinx-doc:
+.. _boundary-conds:
```
Comment (Yurlungur) [inline doc/sphinx/src/boundary_conditions.rst:1]: I think this was an error. The hyperlink never should have been `_sphinx_doc`

File: doc/sphinx/src/boundary_conditions.rst [hunk 2]
```diff
@@ -100,9 +101,19 @@ Other than these requirements, the ``Boundary`` object can do whatever
...
 
-Per package user-defined boundary conditions
---------------------------------------------
+  A per-variable reflecting boundary condition is available, but you
+  must register it manually. To do so, simply call
+  ``app_in->RegisterDefaultReflectingBoundaryConditions()`` and it
+  will be available as a mesh boundary with the name ``reflecting``.
+  The reason manual registration is required is to support custom
+  reflecting boundary conditions in the case where a single variable
+  is used as the state vector.
```
Comment (Yurlungur) [inline doc/sphinx/src/boundary_conditions.rst:112]: As requested by @pgrete in the prior MR.

File: example/boundary_exchange/boundary_exchange_driver.cpp [hunk 3]
```diff
@@ -76,23 +76,23 @@ int main(int argc, char *argv[]) {
                      ar3_t{1.0, 1.0, 1.0});
   // forest_def.AddFace(0, {n[0], n[1], n[3], n[2]}, ar3_t{0.0, 0.0, 0.0},
   // ar3_t{1.0, 1.0, 1.0});
-  forest_def.AddBC(edge_t({n[0], n[1]}), parthenon::BoundaryFlag::outflow);
-  forest_def.AddBC(edge_t({n[0], n[3]}), parthenon::BoundaryFlag::outflow);
+  forest_def.AddBC(edge_t({n[0], n[1]}));
+  forest_def.AddBC(edge_t({n[0], n[3]}));
```
Comment (Yurlungur) [inline example/boundary_exchange/boundary_exchange_driver.cpp:80]: The default flag is now *user* which then uses the boundary function pointer selected in the input deck. Explicitly specifying the boundary flag is thus unnecessary.

File: example/particles/main.cpp [hunk 4]
```diff
@@ -49,12 +49,14 @@ int main(int argc, char *argv[]) {
...
-  pman.app_input->swarm_boundary_conditions[BoundaryFace::inner_x1] = SwarmUserInnerX1;
-  pman.app_input->swarm_boundary_conditions[BoundaryFace::outer_x1] = SwarmUserOuterX1;
+  pman.app_input->RegisterBoundaryCondition(BoundaryFace::inner_x1,
+                                            parthenon::BoundaryFunction::OutflowInnerX1);
+  pman.app_input->RegisterBoundaryCondition(BoundaryFace::outer_x1,
+                                            parthenon::BoundaryFunction::OutflowOuterX1);
+  pman.app_input->RegisterSwarmBoundaryCondition(BoundaryFace::inner_x1,
+                                                 SwarmUserInnerX1);
+  pman.app_input->RegisterSwarmBoundaryCondition(BoundaryFace::outer_x1,
+                                                 SwarmUserOuterX1);
```
Comment (Yurlungur) [inline example/particles/main.cpp:59]: Because I don't specify a name, these get added as "user"

File: src/application_input.hpp [hunk 5]
```diff
@@ -19,17 +19,25 @@
...
+class MeshBlock;
+class MeshBlockApplicationData;
+class OutputParameters;
+class Packages_t;
+class ParameterInput;
+template <typename T>
+class MeshData;
 
-struct ApplicationInput {
+class ApplicationInput {
```
Comment (Yurlungur) [inline src/application_input.hpp:37]: There's more private scope now, so this seemed like an appropriate change.

File: src/mesh/mesh.cpp [hunk 6]
```diff
@@ -1182,4 +1175,29 @@ Mesh::GetLevelsAndLogicalLocationsFlat() const noexcept {
...
+                   pin->GetOrAddString("parthenon/mesh", "ox2_bc", "outflow"),
+                   pin->GetOrAddString("parthenon/mesh", "ix3_bc", "outflow"),
+                   pin->GetOrAddString("parthenon/mesh", "ox3_bc", "outflow")};
+  mesh_swarm_bc_names = {
+      pin->GetOrAddString("parthenon/swarm", "ix1_bc", mesh_bc_names[0]),
+      pin->GetOrAddString("parthenon/swarm", "ox1_bc", mesh_bc_names[1]),
+      pin->GetOrAddString("parthenon/swarm", "ix2_bc", mesh_bc_names[2]),
+      pin->GetOrAddString("parthenon/swarm", "ox2_bc", mesh_bc_names[3]),
+      pin->GetOrAddString("parthenon/swarm", "ix3_bc", mesh_bc_names[4]),
+      pin->GetOrAddString("parthenon/swarm", "ox3_bc", mesh_bc_names[5])};
```
Comment (Yurlungur) [inline src/mesh/mesh.cpp]: @brryan does this seem reasonable to you?

File: src/mesh/mesh.cpp [hunk 7]
```diff
@@ -151,13 +151,8 @@ Mesh::Mesh(ParameterInput *pin, ApplicationInput *app_in, Packages_t &packages,
...
        pin->GetInteger("parthenon/mesh", "nx3") == 1});
-  mesh_bcs = {
-      GetBoundaryFlag(pin->GetOrAddString("parthenon/mesh", "ix1_bc", "reflecting")),
-      GetBoundaryFlag(pin->GetOrAddString("parthenon/mesh", "ox1_bc", "reflecting")),
-      GetBoundaryFlag(pin->GetOrAddString("parthenon/mesh", "ix2_bc", "reflecting")),
-      GetBoundaryFlag(pin->GetOrAddString("parthenon/mesh", "ox2_bc", "reflecting")),
-      GetBoundaryFlag(pin->GetOrAddString("parthenon/mesh", "ix3_bc", "reflecting")),
-      GetBoundaryFlag(pin->GetOrAddString("parthenon/mesh", "ox3_bc", "reflecting"))};
+  SetBCNames_(pin);
+  mesh_bcs = GetBCsFromNames_(mesh_bc_names);
```
Comment (Yurlungur) [inline src/mesh/mesh.cpp:155]: The `mesh_bcs` logic is repeated in the initializer list for each constructor. This is silly. I moved it to a standalone function.

File: scripts/python/packages/parthenon_tools/parthenon_tools/phdf_diff.py [hunk 8]
```diff
@@ -211,14 +211,17 @@ def compare_metadata(f0, f1, quiet=False, one=False, check_input=False, tol=1.0e
...
+    # may vary safely
     f0_Info = {
         key: value
         for key, value in f0.Info.items()
         if key != "Time"
         and key != "BlocksPerPE"
         and key != "WallTime"
         and key != "OutputFormatVersion"
+        and key != "BoundaryConditions"
+        and key != "SwarmBoundaryConditions"
```
Comment (Yurlungur) [inline scripts/python/packages/parthenon_tools/parthenon_tools/phdf_diff.py:224]: This is safe to exclude from metadata checks becuase it was always unused by the infrastructure and this MR (dramatically) changes what this output looks like.

**Comment by Yurlungur** [comment]
I don't understand what's going on with the mac OS CI. It seems to be not properly pulling down the gold files. @pgrete do you have any idea what's going on there?

File: doc/sphinx/src/boundary_conditions.rst [hunk 9]
```diff
@@ -58,13 +59,13 @@ where ``BoundaryFace`` is an enum defined in ``defs.hpp`` as
...
-You can then set this boundary condition via the ``user`` flag in the
-input file:
+You can then set this boundary condition by using the name you
+registered in the input file:
 
 ::
 
    <parthenon/mesh>
-   ix1_bc = user
+   ix1_bc = my_bc_name
```
Comment (pdmullen) [inline doc/sphinx/src/boundary_conditions.rst:68]: Cool 😸

File: src/mesh/mesh.cpp [hunk 10]
```diff
@@ -1182,4 +1175,44 @@ Mesh::GetLevelsAndLogicalLocationsFlat() const noexcept {
...
+                   pin->GetOrAddString("parthenon/mesh", "ix2_bc", "outflow"),
+                   pin->GetOrAddString("parthenon/mesh", "ox2_bc", "outflow"),
+                   pin->GetOrAddString("parthenon/mesh", "ix3_bc", "outflow"),
+                   pin->GetOrAddString("parthenon/mesh", "ox3_bc", "outflow")};
+  // JMM: This is needed because not all BCs are necessarily
+  // implemented for swarms
+  auto maybe = [](const std::string &s) {
+    return ((s == "outflow") || (s == "periodic")) ? s : "outflow";
+  };
+  mesh_swarm_bc_names = {
```
Comment (pdmullen) [inline src/mesh/mesh.cpp:1191]: Cool.  I think this API is more flexible when needing to specify different boundary conditions for fields vs swarms.

Just for my understanding, can you help me work through the scenario that I want to specify user defined boundary conditions that are unique for swarms and fields.  I would do

```
parthenon/mesh/ix1_bc=custom_mesh_ix1
parthenon/swarm/ix1_bc=custom_swarm_ix1
```
with 
```
pman.app_input->RegisterBoundaryCondition(parthenon::BoundaryFace::inner_x1, "custom_mesh_ix1", MyMeshBoundaryInnerX1);
pman.app_input->RegisterSwarmBoundaryCondition(parthenon::BoundaryFace::inner_x1, "custom_swarm_ix1", MySwarmBoundaryInnerX1);
```
 
 or equivalently
 
 ```
parthenon/mesh/ix1_bc=user
```
with 
```
pman.app_input->RegisterBoundaryCondition(parthenon::BoundaryFace::inner_x1, MyMeshBoundaryInnerX1);
pman.app_input->RegisterSwarmBoundaryCondition(parthenon::BoundaryFace::inner_x1, MySwarmBoundaryInnerX1);
```

is that right?

**Review by pdmullen — APPROVED** [review]

File: src/mesh/mesh.cpp (hunk 10 above)
Comment (Yurlungur) [reply src/mesh/mesh.cpp:1191]: Yep that's right!

**Review by lroberts36 — APPROVED** [review]
LGTM

**Pull request merged**